fc_util/src/arg_parser.rs

```diff
@@ -0,0 +1,581 @@
+}
+
+#[derive(Debug, Default, PartialEq)]
+pub struct ArgParser<'a> {
+    args_matches: HashSet<ArgInfo<'a>>,
+    extra_args: Vec<String>,
+}
+
+impl<'a> ArgParser<'a> {
+    pub fn arg(mut self, arg_info: ArgInfo<'a>) -> Self {
+        self.args_matches.insert(arg_info);
+        self
+    }
+
+    pub fn value(&self, arg_name: &'static str) -> Option<String> {
+        self.args_matches.get(arg_name).unwrap().user_value.clone()
+    }
+
+    pub fn extra_args(self) -> Vec<String> {
+        self.extra_args
+    }
+
+    // Used for a general validation of the arguments provided by user.
+    fn is_valid(&self, argument_info: &ArgInfo<'a>, index: usize, args: &[String]) -> Result<()> {
+        // Check if the argument at `index` position was not provided more than once.
+        if self
+            .args_matches
+            .get(argument_info.name)
+            .unwrap()
+            .user_value
+            .is_some()
+        {
+            return Err(Error::DuplicatedArgument(argument_info.name.to_string()));
+        }
+        // If the argument at `index` position takes a value, check if there was provided
+        // one and only one value for that argument.
+        if argument_info.takes_value {
+            if let Some(arg_value) = args.get(index + 1) {
+                if arg_value.starts_with("--")
+                    && args.contains(&(format!("--{}", argument_info.name)))
+                {
+                    return Err(Error::MissingValue(argument_info.name.to_string()));
+                }
+            } else {
+                return Err(Error::MissingValue(argument_info.name.to_string()));
+            }
+        }
+        Ok(())
+    }
+
+    // Used for checking if user provided an argument at the "right" position relative to the
+    // previous argument.
+    fn next_arg(index: usize, args: &[String]) -> Result<()> {
+        if let Some(arg) = args.get(index) {
+            if !arg.starts_with("--") {
+                return Err(Error::UnexpectedArgument(arg.clone()));
+            }
+        }
+        Ok(())
+    }
+
+    pub fn parse(self) -> Result<Self> {
+        let mut args: Vec<String> = env::args().collect();
+        // Remove the first element of `args` as it should be the binary name.
+        args.remove(0);
+
+        self.get_matches(args)
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(mut self, args: Vec<String>) -> Result<Self> {
+        if args.contains(&"--help".to_string()) {
+            println!("Launch a microvm.");
+            for arg in &self.args_matches {
+                println!("{}", arg.format_help());
+            }
+            process::exit(0)
+        }
+
+        self.extra_args = if args.contains(&"--".to_string()) {
+            let index = args.iter().position(|r| r.as_str() == "--").unwrap();
+            let (_, extra_args) = args.split_at(index + 1);
+            extra_args.to_vec()
+        } else {
+            Vec::new()
+        };
```

Comment:
I would do this in the `parse()` method. And then call `get_matches()` supplying only the slice with the parameters before "--" .

Reply:
I had the impression that this should be the desired behavior. I thought that we shouldn't parse the parameters supplied after double dash. To me it doesn't make much sense to both parse the parameters after double dash and also add them to an internal struct member "extra_args". Looks like hardcoded application logic to me that fixes a specific jailer corner case.

From what I could gather a double dash is used to signify the end of command options, after which only positional parameters are accepted. We should check what clap does in this situation.